File: validation_test.go
package main

import (
	"testing"

	"buf.build/go/protovalidate"
	userv1 "github.com/poi2/protovalidate-poc/go/gen/user/v1"
)

func TestCreateUserRequest_Valid(t *testing.T) {
	validator, err := protovalidate.New()
	if err != nil {
		t.Fatalf("failed to create validator: %v", err)
	}

	tests := []struct {
		name    string
		request *userv1.CreateUserRequest
	}{
		{
			name: "valid request",
			request: &userv1.CreateUserRequest{
				Name:                 "John Doe",
				Email:                "john@example.com",
				Password:             "password123",
				PasswordConfirmation: "password123",
			},
		},
		{
			name: "long name",
			request: &userv1.CreateUserRequest{
				Name:                 "A very long name that is still within the 255 character limit",
				Email:                "test@example.com",
				Password:             "securepass",
				PasswordConfirmation: "securepass",
			},
		},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			if err := validator.Validate(tt.request); err != nil {
				t.Errorf("expected valid request, got error: %v", err)
			}
		})
	}
}

func TestCreateUserRequest_Invalid_SingleField(t *testing.T) {
	validator, err := protovalidate.New()
	if err != nil {
		t.Fatalf("failed to create validator: %v", err)
	}

	tests := []struct {
		name        string
		request     *userv1.CreateUserRequest
		expectedErr string
	}{
		{
			name: "empty name",
			request: &userv1.CreateUserRequest{
				Name:                 "",
				Email:                "john@example.com",
				Password:             "password123",
				PasswordConfirmation: "password123",
			},
			expectedErr: "name",
		},
		{
			name: "name too long",
			request: &userv1.CreateUserRequest{
				Name:                 string(make([]byte, 256)),
				Email:                "john@example.com",
				Password:             "password123",
				PasswordConfirmation: "password123",
			},
			expectedErr: "name",
		},
		{
			name: "invalid email",
			request: &userv1.CreateUserRequest{
				Name:                 "John Doe",
				Email:                "not-an-email",
				Password:             "password123",
				PasswordConfirmation: "password123",
			},
			expectedErr: "email",
		},
		{
			name: "email too long",
			request: &userv1.CreateUserRequest{
				Name:                 "John Doe",
				Email:                string(make([]byte, 256)) + "@example.com",
				Password:             "password123",
				PasswordConfirmation: "password123",
			},
			expectedErr: "email",
		},
		{
			name: "password too short",
			request: &userv1.CreateUserRequest{
				Name:                 "John Doe",
				Email:                "john@example.com",
				Password:             "short",
				PasswordConfirmation: "short",
			},
			expectedErr: "password",
		},
		{
			name: "password too long",
			request: &userv1.CreateUserRequest{
				Name:                 "John Doe",
				Email:                "john@example.com",
				Password:             string(make([]byte, 73)),
				PasswordConfirmation: string(make([]byte, 73)),
			},
			expectedErr: "password",
		},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			err := validator.Validate(tt.request)
			if err == nil {
				t.Error("expected validation error, got nil")
				return
			}

			// Check that error message contains the expected field name
			if !contains(err.Error(), tt.expectedErr) {
				t.Errorf("expected error to contain %q, got: %v", tt.expectedErr, err)
			}
		})
	}
}

func TestCreateUserRequest_Invalid_MultipleFields(t *testing.T) {
	validator, err := protovalidate.New()
	if err != nil {
		t.Fatalf("failed to create validator: %v", err)
	}

	tests := []struct {
		name        string
		request     *userv1.CreateUserRequest
		expectedErr string
	}{
		{
			name: "passwords do not match",
			request: &userv1.CreateUserRequest{
				Name:                 "John Doe",
				Email:                "john@example.com",
				Password:             "password123",
				PasswordConfirmation: "different",
			},
			expectedErr: "passwords must match",
		},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			err := validator.Validate(tt.request)
			if err == nil {
				t.Error("expected validation error, got nil")
				return
			}

			// Check that error contains the validation message
			if !contains(err.Error(), tt.expectedErr) {
				t.Errorf("expected error to contain %q, got: %v", tt.expectedErr, err)
			}
		})
	}
}

func TestValidationError_Structure(t *testing.T) {
	validator, err := protovalidate.New()
	if err != nil {
		t.Fatalf("failed to create validator: %v", err)
	}

	// Test with multiple validation errors
	request := &userv1.CreateUserRequest{
		Name:                 "", // invalid: too short
		Email:                "not-an-email", // invalid: not an email
		Password:             "password123",
		PasswordConfirmation: "different", // invalid: passwords don't match
	}

	err = validator.Validate(request)
	if err == nil {
		t.Fatal("expected validation error, got nil")
	}

	// Verify error structure - protovalidate returns detailed error information
	t.Logf("Validation error structure: %+v", err)
	t.Logf("Error message: %v", err.Error())

	// The error should contain information about multiple violations
	if !contains(err.Error(), "name") {
		t.Error("expected error to mention 'name' field")
	}
	if !contains(err.Error(), "email") {
		t.Error("expected error to mention 'email' field")
	}
}

func contains(s, substr string) bool {
	return len(s) >= len(substr) && (s == substr || len(substr) == 0 || indexOf(s, substr) >= 0)
}

func indexOf(s, substr string) int {
	for i := 0; i <= len(s)-len(substr); i++ {
		if s[i:i+len(substr)] == substr {
			return i
		}
	}
	return -1
}

File: ts/__tests__/validation.test.ts
import { describe, test, expect } from 'vitest';
import { create } from '@bufbuild/protobuf';
import { ProtoValidator, ValidationError } from '../src/validation/validator';
import { CreateUserRequestSchema } from '@gen/user/v1/user_pb';
import type { CreateUserRequest } from '@gen/user/v1/user_pb';

describe('CreateUserRequest Validation', () => {
  const validator = new ProtoValidator();

  describe('Valid requests', () => {
    test('should validate valid request', () => {
      const request = create(CreateUserRequestSchema, {
        name: 'John Doe',
        email: 'john@example.com',
        password: 'password123',
        passwordConfirmation: 'password123',
      });

      expect(() => validator.validate(CreateUserRequestSchema, request)).not.toThrow();
    });

    test('should validate long name within limit', () => {
      const request = create(CreateUserRequestSchema, {
        name: 'A very long name that is still within the 255 character limit',
        email: 'test@example.com',
        password: 'securepass',
        passwordConfirmation: 'securepass',
      });

      expect(() => validator.validate(CreateUserRequestSchema, request)).not.toThrow();
    });
  });

  describe('Single field validation errors', () => {
    test('should fail with empty name', () => {
      const request = create(CreateUserRequestSchema, {
        name: '',
        email: 'john@example.com',
        password: 'password123',
        passwordConfirmation: 'password123',
      });

      expect(() => validator.validate(CreateUserRequestSchema, request)).toThrow(ValidationError);

      try {
        validator.validate(CreateUserRequestSchema, request);
      } catch (error) {
        expect(error).toBeInstanceOf(ValidationError);
        if (error instanceof ValidationError) {
          expect(error.message).toContain('name');
        }
      }
    });

    test('should fail with name too long', () => {
      const request = create(CreateUserRequestSchema, {
        name: 'x'.repeat(256),
        email: 'john@example.com',
        password: 'password123',
        passwordConfirmation: 'password123',
      });

      expect(() => validator.validate(CreateUserRequestSchema, request)).toThrow(ValidationError);
    });

    test('should fail with invalid email', () => {
      const request = create(CreateUserRequestSchema, {
        name: 'John Doe',
        email: 'not-an-email',
        password: 'password123',
        passwordConfirmation: 'password123',
      });

      expect(() => validator.validate(CreateUserRequestSchema, request)).toThrow(ValidationError);

      try {
        validator.validate(CreateUserRequestSchema, request);
      } catch (error) {
        if (error instanceof ValidationError) {
          expect(error.message).toContain('email');
        }
      }
    });

    test('should fail with email too long', () => {
      const request = create(CreateUserRequestSchema, {
        name: 'John Doe',
        email: 'x'.repeat(256) + '@example.com',
        password: 'password123',
        passwordConfirmation: 'password123',
      });

      expect(() => validator.validate(CreateUserRequestSchema, request)).toThrow(ValidationError);
    });

    test('should fail with password too short', () => {
      const request = create(CreateUserRequestSchema, {
        name: 'John Doe',
        email: 'john@example.com',
        password: 'short',
        passwordConfirmation: 'short',
      });

      expect(() => validator.validate(CreateUserRequestSchema, request)).toThrow(ValidationError);

      try {
        validator.validate(CreateUserRequestSchema, request);
      } catch (error) {
        if (error instanceof ValidationError) {
          expect(error.message).toContain('password');
        }
      }
    });

    test('should fail with password too long', () => {
      const request = create(CreateUserRequestSchema, {
        name: 'John Doe',
        email: 'john@example.com',
        password: 'x'.repeat(73),
        passwordConfirmation: 'x'.repeat(73),
      });

      expect(() => validator.validate(CreateUserRequestSchema, request)).toThrow(ValidationError);
    });
  });

  describe('Multiple fields validation', () => {
    test('should fail when passwords do not match', () => {
      const request = create(CreateUserRequestSchema, {
        name: 'John Doe',
        email: 'john@example.com',
        password: 'password123',
        passwordConfirmation: 'different',
      });

      expect(() => validator.validate(CreateUserRequestSchema, request)).toThrow(ValidationError);

      try {
        validator.validate(CreateUserRequestSchema, request);
      } catch (error) {
        if (error instanceof ValidationError) {
          expect(error.message).toContain('passwords must match');
        }
      }
    });
  });

  describe('ValidationError structure', () => {
    test('should contain detailed violations for multiple errors', () => {
      const request = create(CreateUserRequestSchema, {
        name: '',
        email: 'not-an-email',
        password: 'password123',
        passwordConfirmation: 'different',
      });

      try {
        validator.validate(CreateUserRequestSchema, request);
        expect.fail('Expected ValidationError to be thrown');
      } catch (error) {
        expect(error).toBeInstanceOf(ValidationError);

        if (error instanceof ValidationError) {
          // Should have multiple violations
          expect(error.violations.length).toBeGreaterThan(0);

          // Check violation structure
          error.violations.forEach(violation => {
            expect(violation).toHaveProperty('fieldPath');
            expect(violation).toHaveProperty('message');
            expect(violation).toHaveProperty('ruleId');
            expect(violation).toHaveProperty('reasonCode');

            // Reason code should be UPPER_SNAKE_CASE
            expect(violation.reasonCode).toMatch(/^[A-Z][A-Z0-9_]*[A-Z0-9]$/);
          });

          // Error should mention the problematic fields
          expect(error.message.toLowerCase()).toMatch(/name|email/);
        }
      }
    });
  });

  describe('Reason code conversion', () => {
    test('should convert rule_id to reason code', () => {
      const request = create(CreateUserRequestSchema, {
        name: '',
        email: 'john@example.com',
        password: 'password123',
        passwordConfirmation: 'password123',
      });

      try {
        validator.validate(CreateUserRequestSchema, request);
      } catch (error) {
        if (error instanceof ValidationError) {
          const violation = error.violations[0];

          // ruleId should be something like "string.min_len"
          expect(violation.ruleId).toBeTruthy();

          // reasonCode should be UPPER_SNAKE_CASE version
          // e.g., "string.min_len" -> "STRING_MIN_LEN"
          expect(violation.reasonCode).toBe(
            violation.ruleId.replace(/\./g, '_').toUpperCase()
          );
        }
      }
    });
  });
});
